Properly set dependencies for the  libraries

drdanz opened the pull request:
## Build System

* Dependencies for the `INTERFACE` libraries (YARP_cv and YARP_pcl)
  are now properly set (#2438).
* PCL is now required to build the YARP_pcl library.
* OpenCV is now required to build the YARP_cv library.

Fixes #2438

PeterBowman commented:
Thank you for the patch, @drdanz. So far I have only checked and tested YARP_pcl and I confirm it works, but let me add a few remarks:

* I don't think any of these components, especially YARP_pcl, should be disabled if their external dependency (PCL, CV, Eigen) is not found. First of all, it's not harmful at all since those are header-only libraries. Secondly, projects such as mine (https://github.com/robotology/yarp/issues/2438) won't get into trouble anyway thanks to the upstream `find_dependency(PCL)` call propagated to the dowsntream `find_package(YARP_pcl)`: no PCL found on my system = I don't even care if YARP was built with or without PCL support ("built" is a big word anyway, no YARP_pcl binaries are generated). I presume this is mostly relevant to packaged YARP distributions. I also presume you don't prepare your Windows YARP installer to fetch/bundle PCL binaries, which are quite heavy (this beast does even transitively depend on VTK and Boost!). In other words, if you do prepare a YARP distribution with PCL support, it's probably to be expected that YARP installer also installs PCL (+ VTK, Boost...) on the target system. I'd rather pick a lighter YARP installer and check whether any of my projects need to depend on YARP_pcl+PCL, then download and install PCL by myself.
* As you have noticed, `find_package(PCL)` performs a search of all known PCL components and populates a rather long list of targets with their transitive dependencies, including VTK+Boost+FLANN et al., consequently propagated to `INTERFACE_LINK_LIBRARIES` (https://github.com/robotology/yarp/issues/2438#issuecomment-749408873). YARP_pcl only needs pcl_common, pcl_io and their respective dependencies, which can be translated in the following manner:
  ```cmake
  find_package(PCL COMPONENTS common io QUIET) # in YarpFindDependencies.cmake
  list(APPEND YARP_pcl_PUBLIC_DEPS "PCL COMPONENTS common io") # in src/libYARP_pcl/src/CMakeLists.txt
  ```
* This line is buggy:
  https://github.com/robotology/yarp/blob/0481f994c6e03897d038c5f1d1078145646a1772/src/libYARP_pcl/src/CMakeLists.txt#L23
  On PCL 1.9+, you don't need it as PCL started using CMake targets and include directories are thus propagated via `target_link_libraries`. You do have to call this on PCL 1.8 (Ubuntu bionic) and earlier, but note that it should read `${PCL_INCLUDE_DIRS}` instead of `${PCL_INCLUDE_DIR}`. In fact, in the code you pasted at https://github.com/robotology/yarp/issues/2438#issuecomment-749408873 there should be PCL-related stuff in `INTERFACE_INCLUDE_DIRECTORIES`.

PeterBowman commented:
For the record, PCL 1.8 and 1.10 register these (inter)dependencies (from `PCLConfig.cmake`):
* Internally: pcl_io => pcl_common, pcl_octree (=> pcl_common)
* Externally:
  * pcl_common => eigen, boost
  * pcl_io => openni, openni2, pcap, png, vtk, libusb-1.0; additionally PCL 1.8 depends on: ensenso, davidSDK, dssdk

drdanz commented:
@PeterBowman Hi, I'm about to do a minor release, and I'm not sure what to do about this patch, I agree that these components should not be disabled if the libraries are not found, but we do have to support bionic in the 3.4 branch, and in bionic the pcl_common and pcl_io targets do not exist, and therefore the only way is to the variables for now.

I reverted the changes to the YARP_eigen library, since it uses targets, and therefore there are no variables to expand, but both YARP_pcl and YARP_opencv have this issue.

I also changed the dependency for YARP_pcl to "common" and "io" libraries and components only.

I'm not sure if we still have to support bionic for the 3.5 release, but as soon as we drop the support, we can remove this dependency.
We should also check the dependencies for the YARP_cv library.

(Side note: I think that we should also rename the library YARP_opencv for coherence... It's not a "computer vision" library, but a library for compatibility with opencv)

drdanz commented:
Build [22] failed for reasons that are not related to this patch.
I'm merging this, in order to make the release, @PeterBowman feel free to open a new issue (or even better a PR 😄) if you think that something is wrong with this PR.

PeterBowman commented on src/libYARP_pcl/src/CMakeLists.txt:
```diff
@@ -20,9 +20,11 @@ target_compile_features(YARP_pcl INTERFACE cxx_std_14)
-target_include_directories(YARP_pcl SYSTEM INTERFACE ${PCL_INCLUDE_DIR})
-target_link_libraries(YARP_pcl INTERFACE ${PCL_LIBRARIES})
-list(APPEND YARP_pcl_PUBLIC_DEPS PCL)
+target_include_directories(YARP_pcl SYSTEM INTERFACE ${PCL_COMMON_INCLUDE_DIR}
+                                                     ${PCL_IO_INCLUDE_DIR})
```
On PCL 1.10 (using CMake targets), `INTERFACE_INCLUDE_DIRECTORIES` is defined as `"${PCL_${COMPONENT}_INCLUDE_DIRS};${PCL_CONF_INCLUDE_DIR}"`. Not only we are missing the `PCL_CONF_INCLUDE_DIR` part (I don't think this is critical here, anyway), but also note the trailing `S` in the component's include dir variable:
* `${PCL_XXX_INCLUDE_DIR}`: just this component's include header dir(s)
* `${PCL_XXX_INCLUDE_DIRS}`: same as above **plus** dependencies

On PCL 1.8 (Ubuntu bionic), my sample application (see linked issue) throws a compile error when using `PCL_XXX_INCLUDE_DIR` due to an unresolved Eigen dependency.

I think this is better written simply as `${PCL_INCLUDE_DIRS}`, thus accounting for all internal and external dependencies (pcl_io also dependes on pcl_octree).

drdanz replied:
reverted, see b0443ce491b6793746eb8fcb0a586b6e2d2f4b1f

PeterBowman commented on src/libYARP_pcl/src/CMakeLists.txt:
```diff
@@ -20,9 +20,11 @@ target_compile_features(YARP_pcl INTERFACE cxx_std_14)
-list(APPEND YARP_pcl_PUBLIC_DEPS PCL)
+target_include_directories(YARP_pcl SYSTEM INTERFACE ${PCL_COMMON_INCLUDE_DIR}
+                                                     ${PCL_IO_INCLUDE_DIR})
+target_link_libraries(YARP_pcl INTERFACE ${PCL_COMMON_LIBRARIES}
+                                         ${PCL_IO_LIBRARIES})
```
Similarly to the above comment, I'd write `${PCL_LIBRARIES}` even if dealing with PCL 1.9+ and CMake targets. The reason is that on PCL 1.10 neither `${PCL_XXX_LIBRARIES}` nor individual targets deal with 3rd party dependencies. So this is a similar issue to the include dirs counterpart, but here we'll probably see errors on link time.

drdanz replied:
reverted see b0443ce491b6793746eb8fcb0a586b6e2d2f4b1f

PeterBowman reviewed (commented):
Sorry, @drdanz, I was struggling with different scenarios and had not come yet to a complete solution. Dealing with PCL+CMake is a bit cumbersome and I'm afraid adopting modern targets might not be a panacea. This patch is definitely an improvement, though, thank you! I'll try to review this whole issue again later on. I have dropped a few comments for future reference.

drdanz commented on src/libYARP_cv/src/CMakeLists.txt:
```diff
@@ -21,8 +21,8 @@ target_compile_features(YARP_conf INTERFACE cxx_std_14)
 
-target_include_directories(YARP_cv SYSTEM INTERFACE ${OPENCV_INCLUDE_DIR})
-target_link_libraries(YARP_cv INTERFACE ${OPENCV_INCLUDE_DIR})
+target_include_directories(YARP_cv SYSTEM INTERFACE ${OpenCV_INCLUDE_DIR})
```
Thanks for catching this. Fixed in 121da79986ddefe384f01c31deb6d5147706a9d3

drdanz commented:
I agree that modern targets would be better, but unfortunately I can't do it in the stable branch...
Applied @PeterBowman suggestions, thanks for the feedback.
